src/otx/engine/engine.py

```diff
@@ -0,0 +1,407 @@
+"""Module for OTX engine components."""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Iterable
+
+import torch
+from lightning import Trainer, seed_everything
+
+from otx.core.config.device import DeviceConfig
+from otx.core.data.module import OTXDataModule
+from otx.core.model.entity.base import OTXModel
+from otx.core.model.module.base import OTXLitModule
+from otx.core.types.device import DeviceType
+from otx.core.types.task import OTXTaskType
+from otx.core.utils.cache import TrainerArgumentsCache
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+    from lightning import Callback
+    from lightning.pytorch.cli import LRSchedulerCallable, OptimizerCallable
+    from lightning.pytorch.loggers import Logger
+    from lightning.pytorch.utilities.types import EVAL_DATALOADERS
+    from pytorch_lightning.trainer.connectors.accelerator_connector import _PRECISION_INPUT
+
+
+LITMODULE_PER_TASK = {
+    OTXTaskType.MULTI_CLASS_CLS: "otx.core.model.module.classification.OTXMulticlassClsLitModule",
+    OTXTaskType.MULTI_LABEL_CLS: "otx.core.model.module.classification.OTXMultilabelClsLitModule",
+    OTXTaskType.H_LABEL_CLS: "otx.core.model.module.classification.OTXHlabelClsLitModule",
+    OTXTaskType.DETECTION: "otx.core.model.module.detection.OTXDetectionLitModule",
+    OTXTaskType.INSTANCE_SEGMENTATION: "otx.core.model.module.instance_segmentation.OTXInstanceSegLitModule",
+    OTXTaskType.SEMANTIC_SEGMENTATION: "otx.core.model.module.segmentation.OTXSegmentationLitModule",
+    OTXTaskType.ACTION_CLASSIFICATION: "otx.core.model.module.action_classification.OTXActionClsLitModule",
+    OTXTaskType.ACTION_DETECTION: "otx.core.model.module.action_detection.OTXActionDetLitModule",
+}
+
+
+class Engine:
+    """OTX Engine.
+
+    This class defines the common interface for OTX, including methods for training and testing.
+
+    Example:
+        # Auto-Configuration with data_root
+        >>> engine = Engine(
+            data_root=<dataset/path>,
+        )
+        # With Custom OTXModel
+        >>> engine = Engine(
+            data_root=<dataset/path>,
+            model=OTXModel(...),
+            checkpoint=<checkpoint/path>,
+        )
+        # With Custom OTXDataModule
+        >>> engine = Engine(
+            model = OTXModel(...),
+            datamodule = OTXDataModule(...),
+        )
+    """
+
+    def __init__(
+        self,
+        *,
+        data_root: str | Path | None = None,
+        task: OTXTaskType | None = None,
+        work_dir: str | Path = "./otx-workspace",
+        datamodule: OTXDataModule | None = None,
+        model: OTXModel | str | None = None,
```

Comment:
`OTXModel` is a `Torch` model, right? So are we always passing torch model to the Engine instead of lightning model? What if I create a LightningModel and pass it to the engine?

Reply:
LightningModule is not available with the current engine implementation.
Should we make it possible to receive LightningModules?

Reply:
We could make Engine.model have `OTXModel | OTXLitModule`, as shown below.

```python
def __init__(
    ...
    model: OTXModel | OTXLitModule | str | None = None,
    ...
):
    if isinstance(model, OTXLitModule):
        ??? -> optimizer, scheduler update?
        self._model = model
    ...

def train():
    lit_module = self._build_lightning_module()
    lit_module.meta_info = self.datamodule.meta_info

    ...

def _build_lightning_module(
    self,
) -> OTXLitModule:
    if isinstance(self.model, OTXLitModule):
       ??? -> optimizer, scheduler update?
       return self.model
    class_module, class_name = LITMODULE_PER_TASK[self.task].rsplit(".", 1)
    module = __import__(class_module, fromlist=[class_name])
    lightning_module = getattr(module, class_name)
    return lightning_module(
        otx_model=model,
        optimizer=optimizer,
        scheduler=scheduler,
        torch_compile=False,
    )

```
I'm actually confused by ??? part in example code (`_init_` or `_build_lightning_module`). We have a ??? part, should we swap optimizer and scheduler? Or should we use the existing ones?
What do you think?

Reply:
I think we could swap the default optimiser and schduler in LitModel if the user explicitly passes them from the Engine. Otherwise, we could use the default ones that is defined in the lightning module.

@ashwinvaidya17? Any thoughts?

Reply:
As discussed, this otx design assumes that each task would have a dedicated lightning module. However, this may not be in the case in the future. Especially when the migration from mmx is complete.

For example, anomalib has a dedicated lightning module for each model instead of task. This is because, each model might have different behaviour within the lightning hooks. Therefore, anomalib models would not work with this approach

Reply:
Maybe we can pass the torch model as a backbone to the lightning module. Something like this
```python
model = OTXMulticlassClsLitModule(backbone=OTXMobilenet,...)
engine = Engine(model=model, ...)
```

Reply:
I don't think I can say much because it is a work related to OTXLitModule. The current architecture has the entire model, not just the backbone.

Reply:
Alright, let's merge this as is, but we will need to modify this when anomalib is being implemented.